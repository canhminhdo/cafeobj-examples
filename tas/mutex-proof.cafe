in tas.cafe

open TAS .
    ops p q : -> Pid .
    red mutex(init, p, q) .
close

open TAS .
    op s : -> Sys .
    ops p q r : -> Pid .
    -- eq c-enter(s,r) = true .
    eq pc(s, r) = rs .
    eq locked(s) = false .
    eq p = r .
    eq q = r .
    red mutex(s, p, q) implies mutex(enter(s, r), p, q) .
close

open TAS .
    op s : -> Sys .
    ops p q r : -> Pid .
    -- eq c-enter(s,r) = true .
    eq pc(s, r) = rs .
    eq locked(s) = false .
    eq p = r .
    eq (q = r) = false .
    red inv1(s, q) implies mutex(s, p, q) implies mutex(enter(s, r), p, q) .
close

open TAS .
    op s : -> Sys .
    ops p q r : -> Pid .
    -- eq c-enter(s,r) = true .
    eq pc(s, r) = rs .
    eq locked(s) = false .
    eq (p = r) = false .
    eq q = r .
    red inv1(s, p) implies mutex(s, p, q) implies mutex(enter(s, r), p, q) .
close

open TAS .
    op s : -> Sys .
    ops p q r : -> Pid .
    -- eq c-enter(s,r) = true .
    eq pc(s, r) = rs .
    eq locked(s) = false .
    eq (p = r) = false .
    eq (q = r) = false .
    red mutex(s, p, q) implies mutex(enter(s, r), p, q) .
close

open TAS .
    op s : -> Sys .
    ops p q r : -> Pid .
    eq c-enter(s,r) = false .
    red mutex(s, p, q) implies mutex(enter(s, r), p, q) .
close

open TAS .
    op s : -> Sys .
    ops p q r : -> Pid .
    -- eq c-exit(s, r) = true .
    eq pc(s, r) = cs .
    eq p = r .
    red mutex(s, p, q) implies mutex(exit(s, r), p, q) .
close

open TAS .
    op s : -> Sys .
    ops p q r : -> Pid .
    -- eq c-exit(s, r) = true .
    eq pc(s, r) = cs .
    eq (p = r) = false .
    eq q = r .
    red mutex(s, p, q) implies mutex(exit(s, r), p, q) .
close

open TAS .
    op s : -> Sys .
    ops p q r : -> Pid .
    -- eq c-exit(s, r) = true .
    eq pc(s, r) = cs .
    eq (p = r) = false .
    eq (q = r) = false .
    red mutex(s, p, q) implies mutex(exit(s, r), p, q) .
close

open TAS .
    op s : -> Sys .
    ops p q r : -> Pid .
    eq c-exit(s, r) = false .
    red mutex(s, p, q) implies mutex(exit(s, r), p, q) .
close