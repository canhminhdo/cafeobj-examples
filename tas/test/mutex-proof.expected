--> base case
(true):Bool
--> induction case for enter
--> case 1.1.1.1
(true):Bool
--> case 1.1.1.2.1: needs inv1
(true):Bool
--> case 1.1.1.2.2
(true):Bool
--> case 1.1.2.1.1: needs inv1
(true):Bool
--> case 1.1.2.1.2
(true):Bool
--> case 1.1.2.2
(true):Bool
--> case 1.2
(true):Bool
--> case 2
(true):Bool
--> induction case for exit
--> case 1.1
(true):Bool
--> case 1.2.1
(true):Bool
--> case 1.2.2
(true):Bool
--> case 2
(true):Bool
